README.md
# SSOXS
Single Sign On for eXternal Services (SSOXS), a Drupal module

Managing authentication and authorization for virtual communities with a distributed service infrastructure.

<img style="float: right;" src="ssoxs.png">

**Drupal version: 7.x**

This project was developed by [StickyBits][5] (Marc van Dijk) for the [WeNMR project][1].
The project was first deposited on the [developers][2] Drupal [git repository][5] and later duplicated on GitHub.

Licensing: This module is made available under the GNU General Public License (GPL) version 3.0 

Copyright:
  2013, Marc van Dijk, Utrecht University, The Netherlands
  2014, Marc van Dijk, StickyBits, Amersfoort, The Netherlands

Acknowledgement: This module was developed as part of the WeNMR
  project (European FP7 e-Infrastructure grant, contract no. 261572,
  www.wenmr.eu).

**Links:**

* [WeNMR project][1]
* [Marc van Dijk on Drupal][2]
* [SSOXS Drupal project page][3]
* [SSOXS Drupal git repository][4]

---

“Unite and conquer”, professionals are organizing themselves more often in virtual communities. 
These communities link likeminded individuals together enabling a common knowledgebase, professional interaction
and exposure to services valuable to the community. These services are often web-based tools organized through a
common gateway or offered by the community’s users on their own infrastructure using their own implementation.
The “Single Sign On for eXternal Services” module, or SSOXS for short, extends Drupal to serve as a single-sign-on 
authentication, central user management and accounting portal for external web servers or services. 
It provides a complete solutions for all SSO and accounting needs for these communities.

The module offers a “My Services” page on the Drupal users profile page allowing them to easily ‘subscribe’ to 
external services.
The services offered are registered with the module by administrators of the service though the modules 
administration interface. A rich feature set allows the service administrator to fine-tune who may subscribe to
the service, if there are additional subscription requirements to fulfill, the use of Token based access and more.
Registered services are allowed access to the modules API using the flexible XML-RPC communication standard. 
The extendable API enables secure communication between the external service and the SSOXS module enabling not 
only user authentication and authorization but also user management and accounting.
The service administration pages allow the service administrators to manage subscribed users, query the accounting
information provided by the service through the API and get an overview of service usage and performance by 
beautiful reports with interactive graphs.

Finally, the module is designed to be a good “Drupal citizen” extending the core functionality with new features 
like the many other excellent open source modules do. It can therefore directly benefit from the many modules that 
extend Drupal with additional functionality such as authentication and authorization mechanisms like Social media 
credentials, Shibboleth, OAuth and OpenID.
The module even hosts it’s own federate login functionality via the versatile simpleSAMLphp library allowing 
Drupal users to login using federate identity out of the box.

It is the combination between the online SSOXS administration interface and the flexible API that provides a 
closed infrastructure for everything related to authentication, authorization and accounting for a service. 
There is no longer a need to install independent modules or access different online services and “glue” them 
together to achieve the same functionality. The combination of these features makes this module unique. 
The service administrator is in control!

Documentation for the module is included in the modules README.txt file
This module was developed as part of the WeNMR project, and academic e-Infrastructure project funded by the 
European FP7 grant, contract no. 261572, (www.wenmr.eu)


## SSOXS features, a detailed view:
### One time authentication
Allow users to authorize themselves to the external services using their Drupal account credentials. 
The SSOXS module integrates with Drupal’s core user management system and directly benefits from 
the many excellent modules that extend Drupal with additional authentication and authorization 
mechanisms like Social media credentials, SAML, Shibboleth, OAuth and OpenID.

### Federate login enabled
SSOXS offers it’s own federated login facility using the powerful simpleSAMLphp library. 
Identity provides configured in simpleSAMLphp are directly available in the module and can be offered 
on Drupal’s login page. Federated login account management is fully configurable with support for 
account persistence, privacy settings and mapping of attributes provided by the identity providers 
with support for the Profile module.

### Subscription based access
Access to external services is managed using a subscription mechanism. Drupal users can subscribe 
to services through a “My Services” page on their account profile.

### Tailored subscription procedures
No single service is the same. The ssoxs module respects this, offering service administrators with
a rich toolset to fine-tune the subscription procedure to their service. Access to a service can be 
restricted by Drupal role and the subscription procedure supports requirements such as; 
a valid certificate, agreement to license terms, manual subscription approval and the availability of
user profile information.

### Token-based access
The SSOXS module offers token-based access to control the frequency of service use by a subscriber. 
Tokens enable a user to first try a service or allows for implementation of a payment model among others.

### Up on security
The SSOXS module handles potentially sensitive data and does so with care by: offering detailed 
permission settings, support for external SQL databases and BlowFish encrypted API communications. 
The module even comes with a build in database backup functionality.

### Easy service management
Enabling single sign-on for an external service is as easy as filling out one form by an authorized 
Drupal user. The form defines the basic service description, the Drupal users that serve as administrators,
the roles that can subscribe to the service, any signup requirements and the API key to enable secure 
XML-RPC communication between the service and the Drupal site.
After service registration the service administration interfaces created by the ssoxs module allows 
administrators to easily administrate subscribed users, access accounting logs and view service statistics 
as beautiful formatted reports with interactive graphs.

### Easy user management
The service administration pages offer a convenient interface for user management. One interface to edit 
a users subscription, subscribe new users, query the user base with powerful filter options and contact 
your users through an email messaging interface.

### XML-RPC API
Services communicate with the SSOXS module for authentication, authorization and accounting using a 
dedicated API using the XML-RPC protocol. XML-RPC implementations are available for every major programming
language and the module offers ready made program classes in Python and PHP.

### Track your service usage
The XML-RPC API offers functions for service accounting. The accounting information is stored in the 
SSOXS database and made available through the accounting and statistics interfaces. Track access to your
service by user, job status or start/finish time. The accounting interface offers a rich logging facility 
for live logs.
Overall statistics for a service for any give time period are available on the ‘Service statistics’ interface.
Service access statistics including access count, user access and user demographics are displayed as 
interactive graphs.

### Service specific attributes
Services often define specific attributes for a user such as a result URL or an internal id. 
The ssoxs module allows administrators to define variable data fields of the type ‘integer’, ‘selections’ 
or ‘text’ that are stored as part of the user subscription. These attributes can be managed via the user 
management interface and are available via the XML-RPC API.

[1]: http://www.wenmr.eu
[2]: https://www.drupal.org/u/mvdijk
[3]: https://drupal.org/sandbox/mvdijk/2123977
[4]: http://git.drupal.org/sandbox/mvdijk/2123977.git
[5]: http://www.stickybits.nl

CHANGELOG
09-0402-14
----------
- Commit of the initial Drupal 7 version of the ssoxs module. All changes and
  additions listed below.

Drupal site structure changes:
- Module settings change path from /admin/settings/* to /admin/config/*.
  All path occurrences changed.
- Modules need to categorized into groups representing there function.
  SSOXS is grouped with the 'Services' group containing other modules working
  with external services. The menu path of the SSOXS admin pages is changed to
  reflect this grouping from /admin/config/ssoxs/
  to /admin/config/services/ssoxs/
- The file_directory_path method for retrieving the absolute path to the sites
  default 'files' directory has been replaced by the file_default_scheme
  function. Thus get absolute internal path to ssoxs files as:
  drupal_realpath(file_default_scheme() . '://') . '/ssoxs'; - File operations
   are database managed by default in Drupal 7. The ssoxs database backup
   function creates and removes database sql files in an unmanaged fashion.
   Replaced file_delete function with file_unmanaged_delete to enable this.

Database API changes:
- Database manipulation functions such as db_create_table, db_drop_field or
  db_add_field no longer accepts and returns an array with status description
  (.e.a finished=>0). All instances of this array as function input are removed.
- Database db_column_exists function changed to db_field_exists.
- Database API no longer uses INSERT, DELETE and UPDATE calls directly through
  SQL queries but through dedicated objects (e.a. db_insert, db.delete and
  db.update). Updated all.
- The use of db_query("SQL query string") is discouraged in favour of the new
  db_select framework. Convertd all db_query calls.
- Drupal 7 automatically calles hook_schema to create database tables on install
  SSOXS overrides this by first checking for existing ssoxs tables in the active
  database before installing them.
- Safely storing text is now handeled through php PDO database handlers by
  serializing/unserializing the text. No longer a need to use
  mysql_escape_string function.
- The theme layer is greatly simplified in Drupal 7 allowing for less code to
  accomplish the same. Simplified the ssoxs themed pages accordingly.

Compliancy to Drupal coding standards:
- White space: trailing line whitespaces removed, single newlines at end of
  file, all Unix style newlines
- Identation: 2 spaces no tabs
- All binary operators enclosed in two whitespaces
- One space between (type) declaration and its parameter
- Checked all control structures: elseif instead of 'else if', one whitespace
  between control keyword and opening parenthesis.
- Always declare new class instance using the class name including parenthesis.
- All functions have the module name as prefix to avoid namespace collisions.
- All persistent variables have the module name as prefix to avoid namespace
  collisions.

Module functionality changes:
- The Drupal 6 version of the modules asks the service administrator for the
  service URL and IP. The latter is used to restrict XML-RPC access. In the
  Drupal 7 version only the URL is asked and the IP associated to it is
  dynamically fetched when needed by a function in the get_services class.
  This ensures that the URL is the only persistent service identifier, the
  service administrator can easily migrate the service to a different IP without
  the need for an update of the service description.
- Certificate validation. The GRID certificate validation requirement
  implemented in the Drupla 6 version was a WeNMR specifc functionality.
  It required validation by a dedicated external services which also required
  access to the Drupal SSOXS database to register validated certificate
  information. This type of certificate validation is hard to generalize.
  Most certificate validation protocols require the certificate to be installed
  on the same machine and hosted though Apache or similar webservers.
  Such a setup can be involving and usually enables only one certificate at a
  time to be validated which does not allow service specific certificate
  validation. The Drupal 7 SSOXS version implements a simple service specific
  certificate validation mechanism. The service administrator can define the
  URL of an external web service capable of validating the certificate loaded
  in the users browser. Upon service subscription the user is asked to validate
  the certificate by clicking a button. This will trigger a AJAX call to the
  external service on behalf of the user. Return true or false on certificate
  validation. It is simple in the sense that is does not return any additional
  information about the certificate such as expiry date but it's powerful in
  that it enables simple but versatile and service specific certificate
  validation. The only requirement is that the external service implements the
  CORS (Access-Control-Allow-Origin) header definition to allow the AJAX call to
  operate on behalf of the user.
- Token based access. Now extended with a checkbox option to have tokens
  function as a 'try-out' mechanism for the service. In this setup, initial
  subscription by the user will not go to any additional requirement steps but
  simply allow the user to try the service for the number of trials defined by
  the initial token balans. When the trial token balans is depleted the user
  will receive an email with instructions on how to proceed (email message can
  be defined by the administrator). If the user decides to subscribe again, all
  additional requirement steps come into action.
- Add CSS class to SSOXS tables to help themers modify the look of SSOXS data
  views.
- Simplified internal database backup function and improved efficiency.
  Makes daily backups by default until full month of backups have been made.
  Then, save last daily backup as monthly backup and repeat daily backup
  procedure. User can control number of monthly backups to keep till a maximum
  of 120.

js/ssoxs_adminform.js
/**
 * @file
 * Attaches the behaviors service administration page.
 */

(function ($) {

Drupal.behaviors.ssoxsAdminForm = {
  attach: function (context, settings) {

    // jQuery function to fold/unfold individual form fieldsets after clicking a
    // checkbox. Fieldsets are uniquely defined by an id that is passed to the
    // expandSection function by the onclick(expandSection(<unique id>)).
    // Synchronize checking/unchecking when clicking the 'subscribe' button.

    expandSection = function(id) { $('#' + id).slideToggle(); };

    // On window load loop over all 'service-block' classes and check if
    // service checkbox is checked. If so show the license terms if they
    // are available.
    $(window).load(function() {
      $(".service-block").each(function() {
        var block = $(this);
        if ($(this).find(':checkbox').is(':checked')) {
           $license = $(block).find(".license");
           if ($license[0]) {
             $license.show();
           }
         }
      });
    });
  }
};
})(jQuery);


js/ssoxs_myservices.js
/**
 * @file
 * Attaches the behaviors My Services page.
 */

(function ($) {

Drupal.behaviors.ssoxsJobAdmin = {
  attach: function (context, settings) {

    // Perform AJAX call to the certificate validation URL on behalf of the user
    // to check validity of certificate loaded in users browser.
    // The remote server must implement the CORS protocol.
    validateCertificate = function(service, http) {

      // Change Drupal AJAX throbber to spinning
      $("#cert-validate").addClass('active');

      setTimeout(function() {
        var $status = $("#cert-validate");

        var valid = false;
        $.ajax({
           async: false,
           type: 'GET',
           url: http,
           success: function(data) {
             valid = true;
           }
        });

        if (valid) {
          $status.replaceWith('<span id="cert-validate" class="success">Certificate validation successful</span>');
          $("#edit-service-" + service + "-certificate").attr('checked', 'checked');
        }
        else {
          $status.replaceWith('<span id="cert-validate" class="failed">Certificate validation failed</span>');
          $("#edit-service-" + service + "-certificate").removeAttr('checked');
        }
      }, 1000);
    };

    $(window).load(function() {

      // Bind click event to the service name table header cell (class status0).
      $('.name').click(function() {

        var $parent = $(this).closest('table');
        $parent.find('tbody').slideToggle();

      });

      // Bind click event to all table 'step' cells. Click on cell
      // expands re-wrapper div and folds all other ones.
      $('.step').click(function() {

        var target = $(this).parent().get(0);
        var $tbody = $(this).closest('tbody');

        $tbody.find('tr').each(function() {
          if (this === target) {
            $(this).find('.req-wrapper').slideToggle();
          }
          else {
            $checkvis = $(this).find('.req-wrapper');
            if ($checkvis.is(':visible')) {
              $checkvis.slideToggle();
            }
          }
        });

      });
    });
   } 
 };
})(jQuery);


connection_scripts/ssoxs_connect.py
#!/usr/bin/env python
# encoding: utf-8

"""
ssoxs.py

Created by Marc van Dijk on 2014-03-01.
Copyright (c) 2013 StickyBits. All rights reserved (www.stickybits.nl).
"""

import sys, os, hashlib, xmlrpclib, random, string
from Crypto.Cipher import Blowfish
from base64 import urlsafe_b64decode, urlsafe_b64encode
from optparse import *

#SERVICE SPECIFIC VARIABLES
USAGE = ""
CRYPT_KEY = '< your secret key >'
SERVER_URL = '< server url >'

class Blowfish_crypt_decrypt(object):

  # Blowfish encryption decryption class
  # - Data being encrypted should have length divisible by 16, pad by spaces if needed.
  # - Encrypt with mode CBC with random IV of 8 bits
  # - Add 8 characters at the beginning of the string because these 8 are not properly encoded
  #   decoded by Blowfish in mode CBC. As it are always the first 8 bits we can remove them when decoding
  # - Use 56 bits secret key

  def __init__(self, encrypt_method=Blowfish.MODE_CBC):

    self.iv = ''.join(random.choice(string.ascii_lowercase + string.digits) for x in range(8))
    self.cipher = Blowfish.new(CRYPT_KEY, encrypt_method, self.iv)

  def encrypt(self, data):

    data = '########' + data
    while (len(data) % 16 != 0):
      data = data + " "

    return urlsafe_b64encode(self.cipher.encrypt(data))

  def decrypt(self, data):

    decipher = self.cipher.decrypt(urlsafe_b64decode(data))
    decipher = (decipher.strip())[8:]
    return decipher

class ssoxs_connect(object):

	# Class for managing user authentication, automatic login and accounting for managed services
	# Successful authentication will populate the user dictionary with user specific info:
	# - dn: users certificate dn
	# - expd: experation date of the certificate
	# - name: user name
	# - mail: user email
	# - subscription: service subscription status as 'inactive', 'pending', 'active'
	# - ip: if the user is logged into the Drupal website this field will list the users IP
	# - uid: users unique uid as used in de Drupal site
	# - rdate: UNIX time stamp of the subscription start date
	# - any service specific information can be added to the dictionary.

  def __init__(self, service):

    self.server = xmlrpclib.Server(SERVER_URL, allow_none=True)
    self.user = {}
    self.service = service

  def autologin(self, dn, encrypt=False):

    # Authentication by public key of the GRID certificate (DN)
    # - DN is send without encryption by default, encrypt=False.
    # - Returned user object or emtpy array decrypted.
    # - Return True or False for aythentication, register user object in the class 'user' attribute

    if encrypt:
      crypt = Blowfish_crypt_decrypt()
      dn = crypt.encrypt(dn)

    try:
      self.user = self.server.ssoxs.autologin(self.service, dn)
    except xmlrpclib.Error, v:
      sys.stderr.write("ERROR: %s" % v)

    if encrypt:
      for data in self.user:
        self.user[data] = crypt.decrypt(self.user[data]);

    if not len(self.user) or not self.user.has_key('name'):
      return False

    return True

  def authenticate(self, username, password, encrypt=False):

    # Authenticate the user by name and password.
		# - Password needs to be plain text but all info is encrypted using Blowfish. Plain text
    #   password hashed by Drupal on the server side and compaired with stored user password has.
		# - Username and password converted to ',' separated string and encrypted using Blowfish
		# - Encrypted string send to server using XML-RPC authenticate method
		# - Returned user object or emtpy array decrypted.
		# - Return TRUE or FALSE for aythentication, register user object in the class 'user' attribute

    userstring = "%s,%s" % (username, password)
    crypt = Blowfish_crypt_decrypt()
    encrypt_string = crypt.encrypt(userstring)

    try:
      self.user = self.server.ssoxs.authenticate(self.service, encrypt_string)
    except xmlrpclib.Error, v:
      sys.stderr.write("ERROR: %s" % v)

    if encrypt:
      for data in self.user:
        self.user[data] = crypt.decrypt(self.user[data]);

    if not len(self.user) or not self.user.has_key('name'):
      return False

    return True

  def accounting(self, uid=0, ip=None, jid=None, status=0, message='', url=None):

		# Communicate job accounting with the server. Possible attributes:
		# - uid: The users uid as registered in the users table of the website the SSOXS module is installed
		#        The default uid of 0 equals a anonymous user.
		# - ip: The IP address the users request came from.
		# - jid: A unique numerical job identifier. If none supplied the SSOXS module will create one for
		#        internal bookkeeping and return it.
		# - status: The status of the job as number; 0 = submitted, 1 = waiting, 2 = ready, 3 = scheduled,
		#            4 = running, 5 = done, 6 = cleared, 7 = failed. Default status set to 0.
		# - message: A log message for the job. The SSOXS module will handle formatting the message and
		#						 appending to the log.
		# - Return the statistics object.

    try:
      stats = self.server.ssoxs.accounting({'machine_name': self.service,'uid': uid,'ip': ip,'jid': jid,'status': status,'message': message,'url': url})
    except xmlrpclib.Error, v:
      sys.stderr.write("ERROR: %s" % v)

    return stats

  def query(self, key, value=None, encrypt=False):

    # Query the user table of the service for specific information specified by a key/value pair
    # (table column/row). If the data is found, the server will try to return the associated user object.
    # Note: if multiple users are found, self.user will turn into a list
    # Function return True if found False if not.

    found = False
    try:
      if not value:
        self.user = self.server.ssoxs.query({'machine_name': self.service, 'key': "%s" % key})
      else:
        self.user = self.server.ssoxs.query({'machine_name': self.service, 'key': "%s" % key, 'value': "%s" % value})
      if self.user: found = True
    except xmlrpclib.Error, v:
      sys.stderr.write("ERROR: %s" % v)

    if encrypt:
      for data in self.user:
        self.user[data] = crypt.decrypt(self.user[data]);

    if len(self.user) == 1:
      self.user = self.user[0]

    return found

  def update(self, encrypt=False):

    # Update user information. Edit information in the self.user dictionary, the run 'update' function.
    # Note, if user info derived from query run that yielded more than 1 user, only first user is used
    # in update function.

    updated_user = None
    try:
      if type(self.user) == type([]):
        updated_user = self.server.ssoxs.update({'machine_name': self.service, 'user': self.user[0]})
      else:
        updated_user = self.server.ssoxs.update({'machine_name': self.service, 'user': self.user})
    except xmlrpclib.Error, v:
      sys.stderr.write("ERROR: %s" % v)

    if updated_user:
      self.user = updated_user

    return updated_user

class CommandlineOptionParser:

	"""Parses command line arguments using optparse"""

	def __init__(self):

		self.option_dict = {}
		self.option_dict = self.CommandlineOptionParser()

	def CommandlineOptionParser(self):

		"""Parsing command line arguments"""

		usage = "usage: %prog" + USAGE
		parser = OptionParser(usage)

		parser.add_option( "-u", "--uid", action="store", dest="uid", type="int", default=0, help="Unique user ID")
		parser.add_option( "-i", "--ip", action="store", dest="ip", type="string", help="IP where the request originated from")
		parser.add_option( "-j", "--jid", action="store", dest="jid", type="int", help="Unique job identifier")
		parser.add_option( "-s", "--status", action="store", dest="status", type="int", default=0, help="Status of the job")
		parser.add_option( "-m", "--message", action="store", dest="message", default=False, help="Optional, messages for the user")
		parser.add_option( "-r", "--url", action="store", dest="url", help="Optional, define URL to a results or project page")

		(options, args) = parser.parse_args()

		self.option_dict['uid'] = options.uid
		self.option_dict['ip'] = options.ip
		self.option_dict['jid'] = options.jid
		self.option_dict['status'] = options.status
		self.option_dict['message'] = options.message
		self.option_dict['url'] = options.url

		return self.option_dict

if __name__ == '__main__':

  commands = CommandlineOptionParser().option_dict

connection_scripts/ssoxs_connect.php
<?php

/**
* @file
* ssoxs.php
*
* Created by Marc van Dijk on 2014-03-01.
* Copyright (c) 2013 StickyBits. All rights reserved (www.stickybits.nl).
*/

class Blowfish_crypt_decrypt {

  /**
   * Blowfish encryption decryption class.
   *
   * - Supports php arrays by serializing/unserializing.
   * - Data being encrypted should have length divisible by 16,
   *   pad by spaces if needed.
   * - Encrypt with mode ECB.
   * - Use 56 bits secret key.
   */

  const CYPHER = MCRYPT_BLOWFISH;
  const MODE   = MCRYPT_MODE_CBC;
  const KEY    = '< your secret key >';

  public function __construct() {

    // Test if xmlrpc extension is loaded, exit with error if not.
    extension_loaded('mcrypt') or exit("PHP mcrypt extension not loaded");
  }

  protected function urlsafeB64encode($string) {

    return strtr(base64_encode($string), '+/=', '-_~');
  }

  protected function urlsafeB64decode($string) {

    return base64_decode(strtr($string, '-_~', '+/='));
  }

  public function encrypt($data) {

    $data = '########' . $data;
    $data = str_pad($data, ceil(strlen($data) / 16 * 16));

    $cypher = mcrypt_module_open(self::CYPHER, '', self::MODE, '');
    $iv = mcrypt_create_iv(mcrypt_enc_get_iv_size($cypher), MCRYPT_RAND);
    mcrypt_generic_init($cypher, self::KEY, $iv);
    $encrypted_data = mcrypt_generic($cypher, $data);
    mcrypt_generic_deinit($cypher);
    mcrypt_module_close($cypher);

    return self::urlsafeB64encode($encrypted_data);
  }

  public function decrypt($crypttext) {

    $crypttext = self::urlsafeB64decode($crypttext);

    $cypher = mcrypt_module_open(self::CYPHER, '', self::MODE, '');
    $iv = mcrypt_create_iv(mcrypt_enc_get_iv_size($cypher), MCRYPT_RAND);
    mcrypt_generic_init($cypher, self::KEY, $iv);
    $data = mdecrypt_generic($cypher, $crypttext);
    mcrypt_generic_deinit($cypher);
    mcrypt_module_close($cypher);

    $data = substr($data, 8);
    return trim($data);
  }
}

class ssoxs_connect
{

  // Single Sign-On for eXternal Services XML-RPC communication class.
  protected $server = "< server url >";
  public $service = NULL;
  public $user = array();

  public function __construct($service) {

    // Dedicated class constructor. Requires service machine name as argument.
    // Test if xmlrpc extension is loaded, exit with error if not.
    extension_loaded('xmlrpc') or exit("PHP xmlrpc extension not loaded");
    $this->service = $service;
  }

  protected function create_xmlrpc_context($request) {

    // Returns the header for the XMLRPC request message.
    $context = stream_context_create(array(
      'http' => array(
      'method' => "POST",
      'header' => "Content-Type: text/xml\r\nUser-Agent: PHPRPC/1.0\r\n",
      'content' => $request,
    )));

    return $context;
  }

  /**
   * Authentication by public key of the GRID certificate (DN).
   *
   * - DN is send without encryption by default, $encrypt=FALSE.
   * - Returned user object or emtpy array decrypted.
   * - Return TRUE or FALSE for authentication, register user object in the
   *   class 'user' attribute.
   */
  public function autologin($dn, $encrypt=FALSE) {

    if ($encrypt) {
      $crypt = new Blowfish_crypt_decrypt;
      $dn = $crypt->encrypt($dn);
    }

    $request = xmlrpc_encode_request("ssoxs.autologin", array($this->service, $dn));
    $file = file_get_contents($this->server, FALSE, $this->create_xmlrpc_context($request));
    $response = xmlrpc_decode($file);

    if ($encrypt) {
      foreach ($response as $key => $data) {
        $response[$key] = $crypt->decrypt($data);
      }
    }

    $this->user = $response;

    if (empty($response) or !(array_key_exists('name', $response))) {
      return FALSE;
    }
    return TRUE;
  }

  /**
   * Authenticate the user by name and password.
   *
   * - Password always communicated as MD5 hash
   * - Username and password converted to ',' separated string and encrypted
   *   using Blowfish.
   * - Encrypted string send to server using XML-RPC authenticate method
   * - Returned user object or emtpy array decrypted.
   * - Return TRUE or FALSE for aythentication, register user object in the
   *   class 'user' attribute.
   */
  public function authenticate($username, $password, $encrypt=FALSE) {

    $userstring = sprintf("%s,%s", $username, $password);

    $crypt = new Blowfish_crypt_decrypt;
    $encrypt_string = $crypt->encrypt($userstring);

    $request = xmlrpc_encode_request("ssoxs.authenticate", array($this->service, $encrypt_string));
    $file = file_get_contents($this->server, FALSE, $this->create_xmlrpc_context($request));
    $response = xmlrpc_decode($file);

    if ($encrypt) {
      foreach ($response as $key => $data) {
        $response[$key] = $crypt->decrypt($data);
      }
    }

    $this->user = $response;

    if (empty($response) or !(array_key_exists('name', $response))) {
      return FALSE;
    }
    return TRUE;
  }

  /**
   * Communicate job accounting with the server. Possible attributes:
   *
   * - uid: The users uid as registered in the users table of the website the
   *   SSOXS module is installed. The default uid of 0 equals a anonymous user.
   * - ip: The IP address the users request came from.
   * - jid: A unique numerical job identifier. If none supplied the SSOXS
   *   module will create one for internal bookkeeping and return it.
   * - status: The status of the job as number; 0 = submitted, 1 = waiting,
   *   2 = ready, 3 = scheduled, 4 = running, 5 = done, 6 = cleared, 7 = failed.
   *   Default status set to 0.
   * - message: A log message for the job. The SSOXS module will handle
   *   formatting the message and appending to the log.
   * - url: Optional URL of the job results page.
   */
  public function accounting($uid=0, $ip=NULL, $jid=NULL, $status=0, $message='', $url=NULL) {

    $response = array();
    $request = xmlrpc_encode_request("ssoxs.accounting", array('machine_name' => $this->service,
                                                               'uid' => $uid,
                                                               'ip' => $ip,
                                                               'jid' => $jid,
                                                               'status' => $status,
                                                               'message' => $message,
                                                               'url' => $url));
    $file = file_get_contents($this->server, false, $this->create_xmlrpc_context($request));
    $response = xmlrpc_decode($file);

    return $response;
  }

  /**
   * Query the service user table for information specified by a key/value pair.
   *
   * - (table column/row). If the data is found, the server will try to return
   *   the associated user object.
   * - Note: if multiple users are found, self.user will turn into a list.
   * - Function return True if found False if not.
   */
  public function query($key, $value=NULL, $encrypt=FALSE) {

    if (!$value) {
      $request = xmlrpc_encode_request("ssoxs.query", array('machine_name' => $this->service, 'key' => $key));
    }
    else {
      $request = xmlrpc_encode_request("ssoxs.query", array('machine_name' => $this->service, 'key' => $key, 'value' => $value));
    }
    $file = file_get_contents($this->server, FALSE, $this->create_xmlrpc_context($request));
    $response = xmlrpc_decode($file);

    if ($encrypt) {
      $crypt = new Blowfish_crypt_decrypt;
      foreach ($response as $key => $data) {
        $response[$key] = $crypt->decrypt($data);
      }
    }

    $this->user = $response;

    if (empty($response) or !(array_key_exists('name', $response))) {
      return FALSE;
    }
    return TRUE;
  }

  /**
   * Update user information.
   *
   * - Edit information in the self.user dictionary, the run 'update' function.
   * - Note, if user info derived from query run that yielded more than 1 user,
   *   only first user is used in update function.
   */
  public function update($encrypt=FALSE) {

    if ($this->user) {
      $request = xmlrpc_encode_request("ssoxs.update", array('machine_name' => $this->service, 'user' => $this->user));
      $file = file_get_contents($this->server, FALSE, $this->create_xmlrpc_context($request));
      $response = xmlrpc_decode($file);
    }

    if ($encrypt) {
      $crypt = new Blowfish_crypt_decrypt;
      foreach ($response as $key => $data) {
        $response[$key] = $crypt->decrypt($data);
      }
    }

    if ($response) {
      $this->user = $response;
    }

    return $response;
  }
}
